lib/src/repo.rs

```diff
@@ -965,3 +978,14 @@ impl MutableRepo {
         );
     }
 }
+
+/// Error from attempts to check out the root commit for editing
+#[derive(Debug, Copy, Clone, Error)]
+#[error("Opened root commit")]
+pub struct OpenedRootCommit;
```

Comment:
I haven't seen this pattern of fine-grained error reporting elsewhere in the code base, but I'm a fan of it. It's even nicer if jj eventually adopts composable polymorphic errors as in e.g. `anyhow`. Thoughts?

Reply:
I don't think `anyhow` is good to use for a library, where we want to exhaustively describe the kinds of errors that can be produced. I've been using `thiserror` in `git-branchless` for some things, which has been working well.

Reply:
Right, I was thinking more about the CLI there. `thiserror` is in use right here already! For the lib side of things, errors like this get aggregated into appropriate public enums.

Reply:
Yes, fine-grained errors like this is what I've tried to do in the lib crate so far.